src/Repartition.chpl

```diff
@@ -24,6 +27,495 @@ module Repartition
     }
   }
 
+  // This function attempts to grab as much as possible from the local slice of the bytes.
+  // Given a SegString (global offsets + global byte buffer), it figures out for each
+  // locale which *parts* of which strings intersect that locale's local chunk of bytes,
+  // and collects bookkeeping needed to rebuild those pieces into local strings.
+  proc balancedSegStringRetrieval(const ref strings: SegString):
+    [PrivateSpace] innerArray(string)
+  {
+    // References to the global segment offsets and underlying byte array
+    ref offsets  = strings.offsets.a;
+    ref strBytes = strings.values.a;
+
+    // Per-locale first and last index (global) of local chunk; -1 means empty.
+    // These are global byte indices into strBytes for each locale’s localSubdomain.
+    var firstIdxPerLoc: [0..#numLocales] int = -1;
+    var lastIdxPerLoc: [0..#numLocales] int = -1;
+
+    // Per-locale end of string corresponding to last offset (global) of local chunk; -1 means empty.
+    // This will hold, for each locale, the byte index where the *last* string that begins
+    // on this locale ends (exclusive). Used as an upper bound when slicing.
+    var lastOffsetEndPerLoc: [0..#numLocales] int = -1;
+
+    // Last byte index in the global byte buffer
+    const endOfBytes = strBytes.domain.last;
+
+    // For each locale, the first string-start offset (global) that resides on that locale.
+    // -1 means that locale has no offsets.
+    var firstOffsetPerLoc: [0..#numLocales] int = -1;
+
+    // First pass: for each locale, record its local ranges in the global bytes and offsets.
+    forall loc in 0..#numLocales {
```

Comment:
Right now, this loop runs entirely on locale 0 (though in parallel). It looks like this could actually be done in parallel by locales simultaneously, if I am not mistaken. If that's the case, `coforall loc in Locales.domain do on Locales[loc] {` could be a better loop+`on` combo here.